### PR TITLE
Compile nimbus-fml for use in a gradle plugin

### PR DESCRIPTION
**Pull request by jhugman**
### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -451,30 +466,47 @@ jobs:
...
-            GHR_SHA256=c868ef9fc5dd8c8a397b74d84051d83693c42dd59041cb17b66f90f563477249
-            curl -sfSL --retry 5 --retry-delay 10 -O "https://github.com/tcnksm/ghr/releases/download/v0.12.0/${GHR}.zip"
-            echo "${GHR_SHA256} *${GHR}.zip" | shasum -a 256 -c -
-            unzip "${GHR}.zip"
-            ./${GHR}/ghr -replace "${CIRCLE_TAG}" megazords/ios-rust/MozillaRustComponents.xcframework.zip
+            ghr -replace "${CIRCLE_TAG}" megazords/ios-rust/MozillaRustComponents.xcframework.zip
+  build-nimbus-fml:
+    executor: macos
+    steps:
+      - full-checkout
```
Comment (tarikeshaq) [inline .circleci/config.yml:488]: Definitely for a follow up... Probably not be worth it unless the fml takes a long time to compile, which it doesn't now - 5 mins is not bad at all:

We might be able to cut down some CI time by using `sccache`:
```yaml
- restore_cache:
    name: Restore sccache cache
    key: fml-sccache-cache-macos-{{ arch }}
- full-checkout
- install rust
- setup-rust-target-version
- setup-sccache

.. rest of commands
- save_cache:
    name: Save sccache cache
    key: fml-sccache-cache-macos-{{ arch }}-{{ epoch }}
    paths:
      - "~/Library/Caches/Mozilla.sccache"
- sccache-stats-and-logs
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -451,30 +466,47 @@ jobs:
...
+      - install-rust
```
Comment (tarikeshaq) [inline .circleci/config.yml:489]: we should probablly add a 
```yaml
- setup-rust-target-version
```
after the `install-rust` to make sure we are running the latest rust we support in app services for this

File: .circleci/config.yml [hunk 3]
```diff
@@ -451,30 +466,47 @@ jobs:
...
+          path: nimbus-fml.zip
+          destination: nimbus-fml.zip
+  release-nimbus-fml:
+    executor: macos
+    steps:
+      - install-ghr
+      - run:
+          name: Release nimbus-fml archive on Github
+          command: |
+            ghr -replace "${CIRCLE_TAG}" nimbus-fml.zip
```
Comment (tarikeshaq) [inline .circleci/config.yml:509]: should we release a sha256 hash of the zip as a checksum something similar as well? so the consumer can verify against it? (I am guessing the android build process will pull this directly from over the network?)

File: .circleci/config.yml [hunk 4]
```diff
@@ -451,30 +466,47 @@ jobs:
...
+          command: |
+            if [ -z "${CIRCLE_TAG}" ]; then
+              # On non-release builds do not update brew (takes ages).
+              echo 'export HOMEBREW_NO_AUTO_UPDATE=1' >> $BASH_ENV
+            fi
+            ./components/support/nimbus-fml/scripts/build-dist.sh
+      - store_artifacts:
+          name: Store nimbus-fml build archive in workspace
+          path: nimbus-fml.zip
+          destination: nimbus-fml.zip
```
Comment (tarikeshaq) [inline .circleci/config.yml:501]: if we end up releasing checksum, we can also store probably it as an artifact (or just print it, as long as someone debugging can easily get the checksum to run android against a debug build of the fml)

File: CHANGES_UNRELEASED.md [hunk 5]
```diff
@@ -18,3 +18,13 @@ Use the template below to make assigning a version number during the release cut
   - Description of the change with a link to the pull request ([#0000](https://github.com/mozilla/application-services/pull/0000))
 
 -->
+
+### ✨✨ What's New ✨✨
+
+#### ⛅️🔭🔬 Nimbus
+
+- Initial release of the Nimbus Feature Manifest Language tool (`nimbus-fml`).
```
Comment (tarikeshaq) [inline CHANGES_UNRELEASED.md:26]: 🎉 🎉 🚀

File: components/support/nimbus-fml/.cargo/config [hunk 6]
```diff
@@ -0,0 +1,2 @@
+[target.x86_64-unknown-linux-musl]
```
Comment (tarikeshaq) [inline components/support/nimbus-fml/.cargo/config:1]: can we add a comment here whey we need this 😅

File: components/support/nimbus-fml/scripts/build-dist.sh [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+fml_dir=$root_dir/components/support/nimbus-fml
+target_dir=$root_dir/target
+filename=$(basename "$fml_dir")
+
+# But we'll dump the zip file wherever we run the script from.
+dist_file=${filename}.zip
+
+prompt='$'
+echo "## Installing tools for cross compiling"
+install_musl_cross="brew install filosottile/musl-cross/musl-cross"
```
Comment (tarikeshaq) [inline components/support/nimbus-fml/scripts/build-dist.sh:29]: Can you write a comment on why we need this?

I see `x86_64-unknown-linux-musl` in our targets, so I imagine the target requires some native libraries that this provides? And for my own curiosity, why do we need to target `x86_64-unknown-linux-musl`, is it for devs running Linux machines?

File: .circleci/config.yml [hunk 8]
```diff
@@ -451,30 +466,47 @@ jobs:
...
           command: |
-            GHR=ghr_v0.12.0_darwin_amd64
-            GHR_SHA256=c868ef9fc5dd8c8a397b74d84051d83693c42dd59041cb17b66f90f563477249
-            curl -sfSL --retry 5 --retry-delay 10 -O "https://github.com/tcnksm/ghr/releases/download/v0.12.0/${GHR}.zip"
-            echo "${GHR_SHA256} *${GHR}.zip" | shasum -a 256 -c -
-            unzip "${GHR}.zip"
-            ./${GHR}/ghr -replace "${CIRCLE_TAG}" megazords/ios-rust/MozillaRustComponents.xcframework.zip
+            ghr -replace "${CIRCLE_TAG}" megazords/ios-rust/MozillaRustComponents.xcframework.zip
+  build-nimbus-fml:
+    executor: macos
```
Comment (tarikeshaq) [inline .circleci/config.yml:486]: definitely not urgent nit:

I see that we use homebrew, which makes sense why we use a macos runner here! But if we can get rid of that, we can run this on a docker runner, which I think is cheaper 😄 That said, I know we'll start shipping this to iOS devs as well... and I'm not sure if once we do that we'll be forced to use macos anyways 🤷

File: .circleci/config.yml [hunk 9]
```diff
@@ -81,6 +81,21 @@ commands:
             ./rustup-init -y --no-modify-path
             rm rustup-init
             echo 'export PATH="$HOME/.cargo/bin:$PATH"' >> $BASH_ENV
+  install-ghr:
```
Comment (tarikeshaq) [inline .circleci/config.yml:84]: Love the refactoring here 😍

**Review by tarikeshaq — APPROVED** [review]
Tried it locally and it works! 🎉 

Dropped a few thoughts, nothing blocking - if anything just a few follow-ups that would be nice

File: components/support/nimbus-fml/scripts/build-dist.sh (hunk 7 above)
Comment (jhugman) [reply components/support/nimbus-fml/scripts/build-dist.sh:29]: We have devs running linux, and Android CI builds are (IIRC) all done on linux.

File: .circleci/config.yml (hunk 8 above)
Comment (jhugman) [reply .circleci/config.yml:486]: That was my thought, yes. Given we're already using homebrew and GHR from a macos runner, and I'm running on macos; I don't know what the M1 cross compilation story is like from linux, etc I went with the simplest option.

File: components/support/nimbus-fml/scripts/build-dist.sh (hunk 7 above)
Comment (tarikeshaq) [reply components/support/nimbus-fml/scripts/build-dist.sh:29]: Aaah the CI builds of course 💯

File: .circleci/config.yml (hunk 1 above)
Comment (jhugman) [reply .circleci/config.yml:488]: Ok, I think I'll leave this for a followup.

File: components/support/nimbus-fml/scripts/build-dist.sh (hunk 7 above)
Comment (jhugman) [reply components/support/nimbus-fml/scripts/build-dist.sh:29]: I've added a comment to explain all.

**Pull request merged**